server/src/main/java/com/octopus/teamcity/opentelemetry/server/OTELHelper.java

```diff
@@ -0,0 +1,101 @@
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.ContextPropagators;
+import io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter;
+import io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporterBuilder;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.sdk.trace.SdkTracerProvider;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
+import io.opentelemetry.sdk.trace.export.SpanExporter;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+import jetbrains.buildServer.log.Loggers;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class OTELHelper {
+
+    private final OpenTelemetry openTelemetry;
+    private final Tracer tracer;
+    private final HashMap<String, Span> spanMap;
+
+    public OTELHelper(Map<String, String> headers, String exporterEndpoint) {
+
+        SpanProcessor spanProcessor = buildSpanProcessor(headers, exporterEndpoint);
+
+        Resource serviceNameResource = Resource
+                .create(Attributes.of(ResourceAttributes.SERVICE_NAME, PluginConstants.SERVICE_NAME));
+        SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
+                .setResource(Resource.getDefault().merge(serviceNameResource))
+                .addSpanProcessor(spanProcessor)
+                .build();
+        this.openTelemetry = OpenTelemetrySdk.builder()
+                .setTracerProvider(sdkTracerProvider)
+                .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
+                .buildAndRegisterGlobal();
+        Loggers.SERVER.info("OTEL_PLUGIN: OTEL_PLUGIN: OpenTelemetry plugin started.");
+        this.tracer = this.openTelemetry.getTracer(PluginConstants.TRACER_INSTRUMENTATION_NAME);
+        this.spanMap = new HashMap<>();
+    }
+
+    private SpanProcessor buildSpanProcessor(Map<String, String> headers, String exporterEndpoint) {
+        OtlpGrpcSpanExporterBuilder spanExporterBuilder = OtlpGrpcSpanExporter.builder();
+        headers.forEach(spanExporterBuilder::addHeader);
+        spanExporterBuilder.setEndpoint(exporterEndpoint);
+        SpanExporter spanExporter = spanExporterBuilder.build();
+
+        Loggers.SERVER.debug("OTEL_PLUGIN: Opentelemetry export headers: " + LogMasker.mask(headers.toString()));
+        Loggers.SERVER.debug("OTEL_PLUGIN: Opentelemetry export endpoint: " + exporterEndpoint);
+
+        return BatchSpanProcessor.builder(spanExporter).build();
+    }
+
+    public boolean pluginReady() {
+        return this.openTelemetry != null;
```

Comment:
```suggestion
    public boolean isReady() {
        return this.openTelemetry != null && this.tracer != null && this.spanMap != null;
```

Reply:
I agree with the this.tracer != null check however the this.spanMap != null was the bug causing the buildListener not to start tracing new spans so will be removed